File: leetcode/two-sum.ts
// twoSum returns indices of two numbers in nums that sum to target.
// Uses a hash map to track complements seen so far in O(n) time.
export function twoSum(nums: number[], target: number): number[] {
  const seen = new Map<number, number>(); // value -> index

  for (let i = 0; i < nums.length; i += 1) {
    const v = nums[i];
    const complementIndex = seen.get(target - v);

    if (complementIndex !== undefined) {
      return [complementIndex, i];
    }

    seen.set(v, i);
  }

  return [];
}

// Demo usage
const nums = [2, 7, 11, 15];
const target = 9;
console.log(twoSum(nums, target)); // [0, 1]

File: leetcode/two-sum.go
package main

import "fmt"

// twoSum finds indices of two numbers in nums that add up to target.
// It uses a hash map to store the complement of each value seen so far.
func twoSum(nums []int, target int) []int {
	seen := make(map[int]int) // value -> index
	for i, v := range nums {
		if j, ok := seen[target-v]; ok {
			return []int{j, i}
		}
		seen[v] = i
	}
	return nil
}

// Demo usage.
func main() {
	nums := []int{2, 7, 11, 15}
	target := 9
	fmt.Println(twoSum(nums, target)) // [0 1]
}
